File: core/src/main/java/com/blockchain/core/fees/FeeDataManager.java
package com.blockchain.core.fees;

import com.blockchain.rx.MainScheduler;

import info.blockchain.wallet.api.FeeApi;
import info.blockchain.wallet.api.data.FeeOptions;
import io.reactivex.rxjava3.core.Observable;

public class FeeDataManager {

    private final FeeApi feeApi;

    public FeeDataManager(FeeApi feeApi) {
        this.feeApi = feeApi;
    }

    /**
     * Returns a {@link FeeOptions} object which contains both a "regular" and a "priority" fee
     * option, both listed in Satoshis/byte.
     *
     * @return An {@link Observable} wrapping a {@link FeeOptions} object
     */
    public Observable<FeeOptions> getBtcFeeOptions() {
        return feeApi.getBtcFeeOptions()
            .onErrorReturnItem(FeeOptions.Companion.defaultForBtc())
            .observeOn(MainScheduler.INSTANCE.main());
    }

    /**
     * Returns a {@link FeeOptions} object which contains both a "regular" and a "priority" fee
     * option for Ethereum.
     *
     * @return An {@link Observable} wrapping a {@link FeeOptions} object
     */
    public Observable<FeeOptions> getEthFeeOptions() {
        return feeApi.getEthFeeOptions()
                .onErrorReturnItem(FeeOptions.Companion.defaultForEth())
                .observeOn(MainScheduler.INSTANCE.main());
    }

    /**
     * Returns a {@link FeeOptions} object which contains both a "regular" and a "priority" fee
     * option for ERC20 tokens.
     * @param contractAddress the contract address for ERC20
     *
     * @return An {@link Observable} wrapping a {@link FeeOptions} object
     */
    public Observable<FeeOptions> getErc20FeeOptions(String parentChain, String contractAddress) {
        return feeApi.getEvmFeeOptions(parentChain, contractAddress)
            .onErrorReturnItem(FeeOptions.Companion.defaultForErc20())
            .observeOn(MainScheduler.INSTANCE.main());
    }

    /**
     * Returns a {@link FeeOptions} object which contains both a "regular" and a "priority" fee
     * option for EVM L1 tokens.
     * @param network the ticker of the native token on the blockchain
     *
     * @return An {@link Observable} wrapping a {@link FeeOptions} object
     */
    public Observable<FeeOptions> getEvmFeeOptions(String network) {
        return feeApi.getEvmFeeOptions(network, null)
            .onErrorReturnItem(FeeOptions.Companion.defaultForEvm(network))
            .observeOn(MainScheduler.INSTANCE.main());
    }

    /**
     * Returns a {@link FeeOptions} object which contains a "regular" fee
     * option, both listed in Satoshis/byte.
     *
     * @return An {@link Observable} wrapping a {@link FeeOptions} object
     */
    public Observable<FeeOptions> getBchFeeOptions() {
        return feeApi.getBchFeeOptions()
                .onErrorReturnItem(FeeOptions.Companion.defaultForBch());
    }

    /**
     * Returns a {@link FeeOptions} object for XLM fees.
     */
    public Observable<FeeOptions> getXlmFeeOptions() {
        return feeApi.getXlmFeeOptions()
                .onErrorReturnItem(FeeOptions.Companion.defaultForXlm());
    }


}

File: core/src/main/java/com/blockchain/core/fees/FeeDataManager.kt
package com.blockchain.core.fees

import com.blockchain.rx.MainScheduler
import info.blockchain.wallet.api.FeeApi
import info.blockchain.wallet.api.data.FeeOptions
import info.blockchain.wallet.api.data.FeeOptions.Companion.defaultForBch
import info.blockchain.wallet.api.data.FeeOptions.Companion.defaultForBtc
import info.blockchain.wallet.api.data.FeeOptions.Companion.defaultForErc20
import info.blockchain.wallet.api.data.FeeOptions.Companion.defaultForEth
import info.blockchain.wallet.api.data.FeeOptions.Companion.defaultForEvm
import info.blockchain.wallet.api.data.FeeOptions.Companion.defaultForXlm
import io.reactivex.rxjava3.core.Observable

class FeeDataManager(private val feeApi: FeeApi) {
    /**
     * Returns a [FeeOptions] object which contains both a "regular" and a "priority" fee
     * option, both listed in Satoshis/byte.
     *
     * @return An [Observable] wrapping a [FeeOptions] object
     */
    val btcFeeOptions: Observable<FeeOptions>
        get() = feeApi.btcFeeOptions
            .onErrorReturnItem(defaultForBtc())
            .observeOn(MainScheduler.main())

    /**
     * Returns a [FeeOptions] object which contains both a "regular" and a "priority" fee
     * option for Ethereum.
     *
     * @return An [Observable] wrapping a [FeeOptions] object
     */
    val ethFeeOptions: Observable<FeeOptions>
        get() = feeApi.ethFeeOptions
            .onErrorReturnItem(defaultForEth())
            .observeOn(MainScheduler.main())

    /**
     * Returns a [FeeOptions] object which contains both a "regular" and a "priority" fee
     * option for ERC20 tokens.
     * @param contractAddress the contract address for ERC20
     *
     * @return An [Observable] wrapping a [FeeOptions] object
     */
    fun getErc20FeeOptions(parentChain: String, contractAddress: String?): Observable<FeeOptions> {
        return feeApi.getEvmFeeOptions(parentChain, contractAddress)
            .onErrorReturnItem(defaultForErc20())
            .observeOn(MainScheduler.main())
    }

    /**
     * Returns a [FeeOptions] object which contains both a "regular" and a "priority" fee
     * option for EVM L1 tokens.
     * @param network the ticker of the native token on the blockchain
     *
     * @return An [Observable] wrapping a [FeeOptions] object
     */
    fun getEvmFeeOptions(network: String): Observable<FeeOptions> {
        return feeApi.getEvmFeeOptions(network, null)
            .onErrorReturnItem(defaultForEvm(network))
            .observeOn(MainScheduler.main())
    }

    /**
     * Returns a [FeeOptions] object which contains a "regular" fee
     * option, both listed in Satoshis/byte.
     *
     * @return An [Observable] wrapping a [FeeOptions] object
     */
    val bchFeeOptions: Observable<FeeOptions>
        get() = feeApi.bchFeeOptions
            .onErrorReturnItem(defaultForBch())

    /**
     * Returns a [FeeOptions] object for XLM fees.
     */
    val xlmFeeOptions: Observable<FeeOptions>
        get() = feeApi.xlmFeeOptions
            .onErrorReturnItem(defaultForXlm())
}
